File: arrayz/easy/second_largest.cpp
#include <iostream>
using namespace std;

// O(N) single pass solution
int secondLargest(int arr[], int n) {
  if (n < 2) {
    return -1;
  }

  int largest = arr[0];
  int second_largest = arr[0];

  for (int i = 0; i < n; i++) {
    if (arr[i] > largest) {
      second_largest = largest;
      largest = arr[i];
    } else if (arr[i] > second_largest && arr[i] != largest) {
      second_largest = arr[i];
    }
  }
  return second_largest;
}

int main() {
  int arr[] = {1, 2, 4, 7, 7, 5};
  int n = sizeof(arr) / sizeof(arr[0]);
  cout << secondLargest(arr, n) << endl;
}

File: arrayz/easy/second_largest.rs
fn second_largest(arr: &[i32]) -> Option<i32> {
    if arr.is_empty() {
        None
    } else {
        let mut largest = arr[0];
        let mut second_largest = arr[0];
        for &item in arr {
            if item > largest {
                second_largest = largest;
                largest = item;
            } else if item > second_largest && item != largest {
                second_largest = item;
            }
        }
        Some(second_largest)
    }
}

fn main() {
    let arr: [i32; 7] = [5, 2, 6, 3, 7, 4, 7];
    match second_largest(&arr) {
        Some(el) => println!("Second largest element is: {}", el),
        None => println!("The array is empty"),
    }
}
